### PR TITLE
Better method completion given syntax errors

### PR DESCRIPTION
**Pull request by jez**
### Motivation

Some of the commits are purely whitespace/clang-format changes. You might want to review by commit or ignore whitespace in the diff.

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by jez** [comment]
- [ ] @jez Add more tests.

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_KwjIDsjKqLJape
→ https://go/builds/bui_KwjIYf2UgAxOdH

**Comment by jez** [comment]
(I don't have any more pending changes to make)

File: third_party/parser/include/ruby_parser/lexer.hh [hunk 1]
```diff
@@ -110,12 +110,12 @@ private:
...
-    std::string tok(const char *start, const char *end);
-    std::string_view tok_view();
-    std::string_view tok_view(const char *start);
-    std::string_view tok_view(const char *start, const char *end);
+    std::string tok() const;
+    std::string tok(const char *start) const;
+    std::string tok(const char *start, const char *end) const;
+    std::string_view tok_view() const;
+    std::string_view tok_view(const char *start) const;
+    std::string_view tok_view(const char *start, const char *end) const;
```
Comment (froydnj) [inline third_party/parser/include/ruby_parser/lexer.hh:118]: Thanks for making these `const`!

File: third_party/parser/include/ruby_parser/lexer.hh [hunk 2]
```diff
@@ -148,8 +148,16 @@ public:
...
 
+    // Main interface consumed by yylex function in parser
     token_t advance();
 
+    // Useful for error recovery. Manually pushes `tok_to_push` (usually: the parser's current
+    // lookahead token) back onto the front of the token_queue, and then allocates a new token and
+    // returns it.
+    token_t unadvance(token_t tok_to_push, token_type type, size_t start, size_t end, const std::string &str);
+
+    std::string_view tok_view(size_t start, size_t end) const;
```
Comment (froydnj) [inline third_party/parser/include/ruby_parser/lexer.hh:159]: Mild preference for calling this `tok_view_from_indices`, though something shorter would be nice.

File: third_party/parser/cc/lexer.rl [hunk 3]
```diff
@@ -177,32 +177,36 @@ void lexer::emit_comment(const char* s, const char* e) {
...
 
-std::string_view lexer::tok_view(const char* start, const char* end) {
+std::string_view lexer::tok_view(const char* start, const char* end) const {
   assert(start <= end);
 
   return std::string_view(start, (size_t)(end - start));
 }
 
+std::string_view lexer::tok_view(size_t start, size_t end) const {
+  return tok_view(source_buffer.data() + start, source_buffer.data() + end);
```
Comment (froydnj) [inline third_party/parser/cc/lexer.rl:207]:
```suggestion
std::string_view lexer::tok_view(size_t start, size_t end) const {
  assert(start < source_buffer.size());
  assert(end <= source_buffer.size());
  return tok_view(source_buffer.data() + start, source_buffer.data() + end);
```

File: third_party/parser/cc/grammars/typedruby.ypp [hunk 4]
```diff
@@ -465,15 +470,63 @@ void parser::error(const std::string &msg) {
...
+	lloc->begin = token->start();
+	lloc->end = token->end();
 	return token_type;
 }
 
+// We've patched the lexer to break compatibility with Ruby w.r.t. method calls
+// for methods sharing names with ruby reserved words. See this PR:
+//   https://github.com/sorbet/sorbet/pull/1993
+// That makes some error recovery better, and other parts worse. This fixes the
+// parts it makes worse.
```
Comment (froydnj) [inline third_party/parser/cc/grammars/typedruby.ypp:488]: This comment is great, thank you for writing it.

**Review by froydnj — APPROVED** [review]
It's nice to see that so few core changes are required for this.

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Kx853iUxskop5a
→ https://go/builds/bui_Kx8667cXBuqMpe

**Comment by jez** [comment]
@froydnj last three commits are new. after rebasing on #5104 i was able to implement some more cases for method completion

File: cfg/builder/builder_walk.cc [hunk 5]
```diff
@@ -293,9 +293,15 @@ BasicBlock *CFGBuilder::walk(CFGContext cctx, ast::ExpressionPtr &what, BasicBlo
 
                 if (a.original) {
                     auto *orig = ast::cast_tree<ast::UnresolvedConstantLit>(a.original);
+                    // Empirically, these are the only two cases we've needed so far to service the
+                    // LSP requests we want (hover and completion), but that doesn't mean these are
+                    // the **only** we'll ever want.
```
Comment (froydnj) [inline cfg/builder/builder_walk.cc:298]:
```suggestion
                    // the **only** cases we'll ever want.
```

**Review by froydnj — APPROVED** [review]

**Pull request merged**